deactivate users

pdurbin opened the pull request:
Note: This feature was originally called "disable users" and was later renamed to "deactivate users" so you will likely see mention of "disable" instead of "deactivate" below.

**What this PR does / why we need it**:

The people who run a Dataverse installation want the ability to deal with users that should be deleted (#4475), disabled (#2419), or otherwise reduced from a normal, active account.

New API endpoints:

- This pull request delivers "disable user account" functionality (via API) as well as documentation in the guides that provides details about the consequences this action. Please note that there is no way to "undisable" an account.
- This pull request also provides a new "Show User Traces" API endpoint. The thinking behind showing the traces is that the "delete" functionality only gives vague messages about why the delete cannot proceed ("the user has created Dataverse object(s)", for example) but now the details are exposed.
- Finally, this pull request includes an API for "remove all roles" that does the equivalent of the button in the superuser dashboard. It was useful in testing (and is included in new tests).

Changes to existing API endpoints:

- The "merge user" API has a new rule: "User accounts can only be merged if they are either both enabled or both disabled." In addition, #7575 ("Unable to merge a user when they have initiated a password reset") was fixed.
- The "delete" API is largely unchanged. An untracked bug was fixed whereby an account can now be deleted even if it has data in the OAuth2TokenData database table. In addition, the delete command now warns that an account cannot be delete if it has been used to create saved searches.
- The "show user as JSON" API is the same except that two new fields have been added: disabled (a boolean), and disabledTime (a timestamp).

**Which issue(s) this PR closes**:

- Closes #2419 User account cannot be disabled/deactivated
- Closes #4475 Delete user via API
- Closes #5462 Password Reset request throws NPE if email is not in authenticateduser table
- Closes #7575 Unable to merge a user when they have initiated a password reset
- Closes #4356 Login: Cannot delete orcid user due to new constraint violation

Note that #4475 was the issue we estimated during sprint planning (large) but as discussion went on we took several turns. We considered the idea that deleting users could mean the database tables would point toward some sort of anonymous or deleted user. We considered promoting merging as a solution for getting rid of users. In the end, we decided to offer the ability to disable a user.

**Special notes for your reviewer**:

We don't have a way to write browser-based tests (#4202) so I didn't include any tests for testing that accounts are disabled via browser/session.

I'm somewhat concerned that we don't include the ability to "undisable" an account but didn't work on this because it doesn't seem strictly necessary and we haven't discussed what that account would be like (permissions, for example, are blown away by the "disable" command). Perhaps this could be added in a future pull request, especially if there's demand for it. In practice it means that if you disable the account of someone using HarvardKey, for example, that person can no longer use HarvardKey to log into a Dataverse installation.

I added the ability to disable a user by both "identifier" (username) and database id because I was looking at the "delete" user code but I'm not sure if this needed or not. We could probably get away with just delete by username.

If you're interested in more discussion on the guides (which served as a spec), see pull request #7585. The branch behind that pull request was eventually merged into this one.

**Suggestions on how to test this**:

First, I tested what I could locally by mocking Shib, and OAuth, but I recommend testing these with real systems. I did not test OIDC accounts at all (I don't know how to) but from talking to Oliver it seems like it should take the same code path as OAuth.

Generally speaking, the guides should be helpful in understanding the expected behavior. See especially bullets under "Disabling a user with this endpoint will:..." One note on "Disabling a user with this endpoint will keep:..." is that disabled users should still show up as contributors under the versions tab of a dataset. I imagine the bulk of the testing involves investigating each of those bullets.

As mentioned above, there's a new rule with regard to merging accounts: "User accounts can only be merged if they are either both enabled or both disabled." The pull request includes automated tests for this but it wouldn't hurt to test it.

As mentioned above, there are two specific bugs that were fixed:

- Closes #5462 Password Reset request throws NPE if email is not in authenticateduser table
- Closes #7575 Unable to merge a user when they have initiated a password reset

**Does this PR introduce a user interface change? If mockups are available, please link/include them here**:

Yes, when accounts are disabled, users will see "Sorry, your account has been disabled. If you believe this is an error, please contact support for assistance." There is a screenshot below. The message should be identical for Builtin, Shib, OAuth, and OIDC accounts.

<img width="1265" alt="account-disabled" src="https://user-images.githubusercontent.com/21006/108557562-c8082480-72c6-11eb-9129-4a71eb1f0c59.png">


If an OAuth account attempts the "convert" process on a disabled account, it should fail with "Your account can only be converted if you provide the correct username and password for your existing account. If your existing account has been disabled by an administrator, you cannot convert your account." The screenshot below shows the old message before it was updated.

<img width="1265" alt="oauth" src="https://user-images.githubusercontent.com/21006/108557541-c2124380-72c6-11eb-9d00-4f88f744c6ed.png">


If a Shib/Institutional account attempts the "convert" process on a disabled account, it should fail with "Validation Error - Your account can only be converted if you provide the correct password for your existing account. If your existing account has been disabled by an administrator, you cannot convert your account." The screenshot below shows the old message before it was updated.

<img width="1242" alt="shib" src="https://user-images.githubusercontent.com/21006/108557522-baeb3580-72c6-11eb-8cc3-1fdb1a1aa9ea.png">


**Is there a release notes update needed for this change?**:

Yes, included.

**Additional documentation**:

None.

sekmiller commented on src/main/java/edu/harvard/iq/dataverse/api/Admin.java:
```diff
@@ -1686,7 +1714,7 @@ public Response submitDatasetVersionToArchive(@PathParam("id") String dsid, @Pat
 			// the API Token which is then used to retrieve files (e.g. via S3 direct
 			// downloads) to create the Bag
-            session.setUser(au);
+            session.setUser(au); // TODO: Stop using session. Use createDataverseRequest instead.
```
Is this still a TODO?

pdurbin replied:
This is a future TODO that I discussed with @qqmyers . I didn't want to mess with this SubmitToArchive (and ask for QA to re-test it) but it's somewhat of session.setUser, I feel. As I indicated, createDataverseRequest is what we usually do in this case.

poikilotherm commented on doc/sphinx-guides/source/api/native-api.rst:
```diff
@@ -3075,9 +3079,104 @@ Deletes an ``AuthenticatedUser`` whose ``id``  is passed. ::
+  export SERVER_URL=http://localhost:8080
+  export USERNAME=jdoe
+
+  curl -H "X-Dataverse-key:$API_TOKEN" -X POST $SERVER_URL/api/admin/authenticatedUsers/$USERNAME/disable
```
Why would it be necessary to use a token here? IIRC all other endpoints in admin API are used without these tokens. For security, you can either use it on localhost only or add the `?unblock-key=...` feature. This might cause confusion.

pdurbin replied:
It's because DisableUserCommand calls RevokeAllRolesCommand, which requires superuser permissions.

scolapasta replied:
We should think of a strategy here to not require a token for an admin api, as I agree it's confusing. Generally (I think this is true), things in the admin API do not call commands, as commands require users. I think that has been a good model (or we need to decide on having commands that don't require users, but that feels dangerous).

pdurbin replied:
Fixed in d6e11732c (using authSvc.getAdminUser as suggested elsewhere).

sekmiller commented on src/main/resources/db/migration/V5.3.0.3__4475-disable-users.sql:
```diff
@@ -0,0 +1,4 @@
+-- Users can be disabled.
```
There is another script named V5.3.0.3 for expanded dataset field validation. How do we go about resolving? (Having deployed the other I'm getting a flyway script error when I try to deploy this branch.)

pdurbin replied:
I'm happy to let @scolapasta make a ruling.

scolapasta replied:
Since expanded dataset field validation is already in QA, let's change this one to be after.

pdurbin replied:
Ok, changed to be after in e73c43b70.

sekmiller reviewed (approved):
I don't have any objections. I'm a little embarrassed that I spelled Authenticated wrong in a method name that is being carried forward in perpetuity, and there may be something to be done about removing groups and group role assignments in tandem, but those are quibbles.

scolapasta commented on src/main/java/edu/harvard/iq/dataverse/DataFile.java:
```diff
@@ -57,6 +57,10 @@
 @NamedQueries({
 	@NamedQuery( name="DataFile.removeFromDatasetVersion",
 		query="DELETE FROM FileMetadata f WHERE f.datasetVersion.id=:versionId and f.dataFile.id=:fileId"),
+        @NamedQuery(name = "DataFile.findByCreatorId",
```
I noticed these are repeated for all the dvobjects? Could we simplify by just getting from dvobject? (is there a case for individually? it just seems a little over kill)

scolapasta replied:
ok.

scolapasta commented on src/main/java/edu/harvard/iq/dataverse/DataverseSession.java:
```diff
@@ -210,6 +221,9 @@ public void dismissMessage(BannerMessage message){
     }
     
     public void configureSessionTimeout() {
+        if (user instanceof GuestUser) {
```
Could you explain what this adds?

pdurbin replied:
Right. This is the thing @landreev and I talked about in Slack: https://iqss.slack.com/archives/C010LA04BCG/p1613509511062000

In short, when `configureSessionTimeout` was added, it was always put in after `setUser`, with the implication that we're now dealing with an AuthenticatedUser (or at least a non-GuestUser).

That is to say, all over the code, you'll see these two lines execute one after the other:

```
setUser(user);
configureSessionTimeout();
```

Prior to this pull request `configureSessionTimeout` could assume it has a non-GuestUser and would go ahead an lengthen the session. Now we can't make that assumption since `setUser` returns early if the user is disabled, with the result that the session contains a GuestUser rather than an AuthenticatedUser.

scolapasta replied:
I just checked and see what you're saying; we *always* call configureSessionTimeout after setUser. Why don't we just move that logic into setUser then? (and it would be after the early return). We can then not have to remember to call it if we add a setUser again somewhere else.

pdurbin replied:
@landreev do you have any objection to what @scolapasta is proposing above?

landreev replied:
I don't remember much about that code (or what exactly I was thinking when I implemented it that way). But yes, it looks like it would be possible - to just call it from inside setUser().

landreev replied:
P.S. as long as we make sure it's called from inside this `if` block in setUser(): 
```
FacesContext context = FacesContext.getCurrentInstance();
if(context != null) {
   ...
}
```
... or otherwise in a way that's not going to result in a null pointer if it's called from a place where faces context is unavailable...

pdurbin replied:
Thanks, all. Fixed in 610e9ca32.

scolapasta commented on src/main/java/edu/harvard/iq/dataverse/UserNotificationServiceBean.java:
```diff
@@ -103,6 +103,10 @@ public void sendNotification(AuthenticatedUser dataverseUser, Timestamp sendDate
     }
 
     public void sendNotification(AuthenticatedUser dataverseUser, Timestamp sendDate, Type type, Long objectId, String comment, AuthenticatedUser requestor, boolean isHtmlContent) {
+        if (dataverseUser.isDisabled()) {
```
I don't think we need this: disabled users will have no roles. Let's save a few lines of code.

pdurbin replied:
Agreed. I was just being paranoid, I guess. Removed in 8f08f7b31.

scolapasta commented on src/main/java/edu/harvard/iq/dataverse/authorization/AuthenticationServiceBean.java:
```diff
@@ -215,6 +220,7 @@ public void deleteAuthenticatedUser(Object pk) {
                 em.remove(confirmEmailData);
             }
             userNotificationService.findByUser(user.getId()).forEach(userNotificationService::delete);
+            em.createNativeQuery("delete from OAuth2TokenData where user_id =" + user.getId()).executeUpdate();
```
can you move this to a named query?

scolapasta replied:
and actually maybe better, why not use  a cascade here too?

pdurbin replied:
I need to refresh myself on the proper setup of cascades, but sure.

pdurbin replied:
Fixed in 8afbf5e92.

scolapasta commented on src/main/java/edu/harvard/iq/dataverse/authorization/providers/builtin/BuiltinUserServiceBean.java:
```diff
@@ -79,6 +79,8 @@ public BuiltinUser find(Long pk) {
         final BuiltinUser user = findByUserName(userName);
         if ( user != null ) {
+            // TODO: Consider adding a cascade delete instead.
+            passwordResetService.deleteResetDataByDataverseUser(user);
```
any reason not to do cascade delete now? seems straightforward and less tech debt.

pdurbin replied:
The only reason is that I hadn't set up a cascade in a while. I'll look at some examples and see what I can do.

pdurbin replied:
Fixed in 6e61841cb.

scolapasta commented on src/main/java/edu/harvard/iq/dataverse/authorization/users/User.java:
```diff
@@ -14,4 +14,6 @@
 
     public boolean isSuperuser();
 
+    public boolean isDisabled();
```
fyi, it is now possible to add a default implementation of the method and so not have to override in subclasses if they use the default.

pdurbin replied:
Gotcha. Do you want me to try to make this change for isDisabled? (I'd rather not mess around with isAuthenticated and isSuperuser right now).

scolapasta replied:
Yeah, let's do it. It's not important enough to hold anything up, but should a good tool to use for these kinds of things in the future.

pdurbin replied:
Done in 23b8bf638.

scolapasta reviewed (commented):
Very extensive - nice! Overall made some small comments.

The two bigger things I have questions on are:
• needing a token for an admin API. I don't think we should do this - if you need a token, we should move to a reachable API. If we want to keep in admin we should refactor, probably by moving the code to a shared place for the command and outside. We likely do want this to be a publicly accessibly (for superuser) API someday, but I still think we should have the code reflect it's current way of being, again admin API means don't use commands that require checking user. (this does complicate things a little, because then we'd probably want to add something to the actionlogrecord for this, since we wouldn't get that for free from the Command. I don't know, there's no simple / ideal path, but the current state just seems to be not ideal.
The 2nd is whether we actually want the system to care about disabled users in a way that's different (from a UI perspective) from when you are actually deleted. Obviously it should not allow reusing a username / email, BUT I could see it be that if you try to login and are disabled it just says "user does not exist" as if it had been deleted. (that's in line with when I was suggesting remove rows from authenticated lookup). 

Let's find some time to discuss these things over slack (or possibly a call) and include any interested parties.

scolapasta commented:
> needing a token for an admin API. I don't think we should do this - if you need a token, we should move to a reachable API. If we want to keep in admin we should refactor, probably by moving the code to a shared place for the command and outside. We likely do want this to be a publicly accessibly (for superuser) API someday, but I still think we should have the code reflect it's current way of being, again admin API means don't use commands that require checking user. (this does complicate things a little, because then we'd probably want to add something to the actionlogrecord for this, since we wouldn't get that for free from the Command. I don't know, there's no simple / ideal path, but the current state just seems to be not ideal.

So I've been thinking more about this. There's room for a lot of cleanup here with how admin api should work with commands and if there should be a SystemUser.java class for example. @pdurbin can you verify for me that none of these commands that are called store the user anywhere? (except probably actionLogRecord)?)

At the very least, let's use some code that harvesting uses (thanks @landreev) and not require a token - it calls a method in the AuthenticationServiceBean:
```authSvc.getAdminUser(); ```

(this gets the first by id superuser - which we should eventually make work better, but is fine for now to take a little debt).

pdurbin reviewed (commented):
@scolapasta thanks for the feedback. Here are some comments on it.

pdurbin commented:
> needing a token for an admin API

Fixed:

d6e11732c run disable user command without superuser API token

With regard to where the superuser is stored/logged, I believe it's only in the ActionLogRecord.

> The 2nd is whether we actually want the system to care about disabled users in a way that's different (from a UI perspective) from when you are actually deleted.

I'd consider this some sort of phase 2 or future effort.

Here are some other fixes I put in the branch:

610e9ca32 merge configureSessionTimeout into setUser
8afbf5e92 remove oauth2tokendata with a cascade 
6e61841cb remove password reset data with cascade

I'm putting this back into review.

scolapasta reviewed (approved):
Good changes - I'll approve.

I did notice there are some calls in deleteAuthenticatedUser that do seem like they could also become cascade... they were there before, so we can consider out of scope, but @pdurbin can you take a look and see if this is correct? And if so open up a new issue for that as it could be a future tech debt cleanup.

pdurbin commented:
Sure, I just opened this issue:

-  more cascade delete for deleteAuthenticatedUser #7652

kcondon commented:
Issues found so far:
1. Disable user endpoint doesn't care what permissions api key has but requires a key, otherwise http 405. Thought discussion in pr and that it is an admin api would not require a key.
2. After disabling a user, some users cannot be found in assign roles ui, autocomplete. Also will not allow assigning role to fully typed existing user that is not disabled but doesn't appear in autocomplete list. Not other users do appear in auto complete list and the ones that do not appear do appear in dashboard, users. Last, this does not happen on develop branch.

pdurbin commented:
@kcondon thanks for all the testing! I was able to reproduce almost all of the issue you found and put in various fixes, detailed below.

@scolapasta @landreev please see the note about getUser below. I'm moving this to code review.

> 1. Disable user endpoint doesn't care what permissions api key has but requires a key, otherwise http 405. Thought, based on discussion in pr and that it is an admin api, that it would not require a key.

I can't reproduce this and the automated tests do not use a token. However, the docs were incorrect and showed the use of an API token so I corrected them in abb8be01e.

> 2. After disabling a user, some existing users cannot be found in assign roles ui, autocomplete.

As far as I can tell, autocomplete for assigning roles was completely broken in the branch, due to me botching the query. Fixed in 0efee2810.

> 3. Attempting to remove roles using api endpoint with a key that does not have sufficient permissions gives incomplete/confusing error message: {"status":"ERROR","message":"Unable to revoke all roles: null"}

The error wasn't being bubbled up from the command. Fixed in 3713104c3.

> 4. On manage users page, disabled user still appears, can make a superuser. Suggestion, make authentication type for disabled user, disabled.

As discussed in Slack, we're putting the word "disabled" in parentheses. In addition the JSON output from /api/admin/list-users now contained "deactivated" and "deactivatedTime". This code and the dashboard code are related. Fixed in c2b7b0aaf.

> 5. Disable user if logged in, remains logged in and can edit account info. Update: this happens with delete user too. For disabled case, reenables it but without some stuff, for delete it adds a new account on save!

Fixed in 3e59e70b6. Now the getUser code checks if the user in the session still exists and is not disabled.

Concern has been expressed that modifying getUser like this could result in a performance problem. A quick benchmark indicates that if you navigate to the homepage, sign up for an account, and then create a dataverse, the getUser method is called 264 times.

> 6. Can log in using builtin disabled user account, if that disabled user edited their account info in step 5

Fixed in 3e59e70b6, the same getUser fix mentioned above.

> 7. Can merge the same account into itself with result being user is deleted. In 5.3, throws constraint error, though could use better error msg.

Indeed, the user was deleted. This bug is probably preexisting in released versions. Fixed in 4f81fbde7.

> 8. Convert oauth (github) account link/page with same email as a newly created, enabled builtin account results in 500 error: 

I can't reproduce this. I'm on 4f81fbde7.

> 9. Convert oauth (github) account with same email as disabled builtin account, sends user to convert account form, then when enter account to convert, provides both a success and failure message and oauth is logged in. Checking account status after shows merge didn't happen.

Fixed in fbded77e1 to prevent deactivated accounts from being converted to OAuth accounts.

> 10. Convert shib account works with enabled builtin but on disabled builtin, gives both success and failure though it actually fails. Should check and prevent submitting account for conversion as ui above.

Fix in fdc9cbe44 to prevent deactivated accounts from being converted to Shib accounts.

scolapasta commented:
> Concern has been expressed that modifying getUser like this could result in a performance problem. A quick benchmark indicates that if you navigate to the homepage, sign up for an account, and then create a dataverse, the getUser method is called 264 times.

This is what I was worried about, 264 queries for this is not great. What about this - getUser(boolean) (and getUser() calls that with false). With true it calls the query, with false just returns the user. And that way you can target its use to only the places where we need to protect for this - user actions, in the command engine when checking permissions, etc.

pdurbin commented:
@scolapasta sure, implemented in d0191c3

We now call the more expensive version if the user tries to go to the account page (which logs them out) of if the user tries to execute a command.

The command I played with the most was "create dataverse" since in prod and most of our dev environments, all authenticated users can create dataverses in root. Here's how the the error looks (the only difference for deleted users is that it says "deleted" instead of "deactivated"):

<img width="1221" alt="Screen Shot 2021-03-18 at 1 50 26 PM" src="https://user-images.githubusercontent.com/21006/111674605-92395b80-87f2-11eb-9f0f-24e11f576b37.png">

pdurbin commented:
The post at https://groups.google.com/g/dataverse-dev/c/wh3m55EGOPU/m/Mij-l0SlAQAJ just reminded me about #4356 about oauth2tokendata which has also been fixed by this pull request (mostly recently touched in 8afbf5e). So I added it to the list of issues above that this pull request closes.

scolapasta commented on src/main/java/edu/harvard/iq/dataverse/EjbDataverseEngine.java:
```diff
@@ -215,7 +218,21 @@ private DvObject getRetType(Object r){
 
             DataverseRequest dvReq = aCommand.getRequest();
-            
+
+            AuthenticatedUser authenticatedUser = dvReq.getAuthenticatedUser();
+            if (authenticatedUser != null) {
+                AuthenticatedUser auFreshLookup = authentication.findByID(authenticatedUser.getId());
+                if (auFreshLookup == null) {
+                    logger.fine("submit method found user no longer exists (was deleted).");
+                    throw new CommandException(BundleUtil.getStringFromBundle("command.exception.user.deleted", Arrays.asList(aCommand.getClass().getSimpleName())), aCommand);
+                } else {
+                    if (auFreshLookup.isDeactivated()) {
+                        logger.fine("submit method found user is deactivated.");
+                        throw new CommandException(BundleUtil.getStringFromBundle("command.exception.user.deactivated", Arrays.asList(aCommand.getClass().getSimpleName())), aCommand);
+                    }
+                }
+            }
+
```
I think we can simplify and remove this - there is already a DataverseRequest associated with this call to the command and that should be enough, as long as we also change the one place where it gets the user from the session in DataverseRequestService to call the new getUser(true) verson of that method.
```
   @PostConstruct
    protected void setup() {
        dataverseRequest = new DataverseRequest(dataverseSessionSvc.getUser(), request);
    }
```

pdurbin replied:
I tried putting it there in the postconstruct and couldn't get it to work. So I put it in the engine instead.

pdurbin replied:
I don't remember a error. I think it simply didn't work. It had no effect, as I recall.

scolapasta replied:
That's weird - since it should set the request to point to a GuestUser in that case and therefore fail due to permission. I would prefer this simpler mechanism if we can get it to work, though (and at least understand why it doesn't if it doesn't). Let me go ahead and try it and see what happens.

pdurbin replied:
Ok. Good luck.

scolapasta reviewed (changes requested):
One suggested change related to getUser (which can simplify the code) and a few other minor questions / notes that may or may not need to be acted on. @pdurbin can you review and let me know what you think?

pdurbin commented on doc/sphinx-guides/source/api/native-api.rst:
```diff
@@ -3055,6 +3055,8 @@ Example: ``curl -H "X-Dataverse-key: $API_TOKEN" -X POST http://demo.dataverse.o
 
 This action moves account data from jsmith2 into the account jsmith and deletes the account of jsmith2.
 
+Note: User accounts can only be merged if they are either both non-deactivated or both deactivated. See :ref:`deactivate a user<deactivate-a-user>`.
```
"Active" is the opposite of "inactive". Is "active" also the opposite of "deactivated"? Maybe. I felt that "non-deactivated" is more accurate.

scolapasta replied:
I agree that "Active" is the opposite of "inactive"; I think of deactivated accounts as inactive, it's just more specific in that it states that they were "deactivated" (how they got to inactive). We probably could have called them "inactive" as well. But changing that now is overkill.

Non-deactivated is wordy and not a good way to describe the the normal state of a user, e.g when you just create one. "I just created an account on Dataverse; it is non-deactivated"

pdurbin replied:
I still prefer the precision but I made the change in df32ec012.

pdurbin commented on src/main/java/edu/harvard/iq/dataverse/DataversePage.java:
```diff
@@ -695,7 +695,7 @@ public String save() {
         } catch (CommandException ex) {
             logger.log(Level.SEVERE, "Unexpected Exception calling dataverse command", ex);
-            String errMsg = create ? BundleUtil.getStringFromBundle("dataverse.create.failure") : BundleUtil.getStringFromBundle("dataverse.update.failure");
+            String errMsg = create ? ex.getLocalizedMessage() : BundleUtil.getStringFromBundle("dataverse.update.failure");
```
Please see d0191c394 for how ex.localized message comes from EjbDataverseEngine with (for example) ` command.exception.user.deactivated={0} failed: User account has been deactivated.`

scolapasta replied:
Ah, I see, but a) I also requested simplifying that code and changing it - and it still isn't clear to me why this is only on "create".

pdurbin replied:
Because "create" is the action that all authenticated users can take in the root (for prod and various dev environments).

pdurbin commented on src/main/java/edu/harvard/iq/dataverse/api/Admin.java:
```diff
@@ -663,6 +695,10 @@ public Response builtin2shib(String content) {
 						password);
 				if (authenticatedUser != null) {
 					knowsExistingPassword = true;
+                                        if (builtInUserToConvert.isDeactivated()) {
```
You're right. It could be. I'm testing moving it up.

pdurbin replied:
Moved up in 530dab34d.

pdurbin commented on src/main/java/edu/harvard/iq/dataverse/engine/command/impl/MergeInAccountCommand.java:
```diff
@@ -57,7 +58,15 @@ public MergeInAccountCommand(DataverseRequest createDataverseRequest, Authentica
+        }
+
+        if (consumedAU.isDeactivated() && !ongoingAU.isDeactivated() || !consumedAU.isDeactivated() && ongoingAU.isDeactivated()) {
+            throw new IllegalCommandException("User accounts can only be merged if they are either both non-deactivated or both deactivated.", this);
```
Please see above.

kcondon commented:
Original list:
Issues found so far:

X Disable user endpoint doesn't care what permissions api key has but requires a key, otherwise http 405. Thought, based on discussion in pr and that it is an admin api, that it would not require a key.

X After disabling a user, some existing users cannot be found in assign roles ui, autocomplete. Also will not allow assigning role to fully typed existing user that is not disabled but doesn't appear in autocomplete list. Note other users do appear in auto complete list and the ones that do not appear do appear in dashboard, users. This does not happen on develop branch. Was able to reproduce if users were first created on develop branch and then the last create was disabled. None of the other users appear in assign roles autocomplete. Note that it does not happen if users were created in the disable user branch.
Steps to reproduce:
-on develop branch, create 4 users with similar names: yak1234ab, yak1234bc, yak1234cd, yak1234de.
-on disable user branch, disable user yak1234de
-as dataverse admin, go to dataset permissions page for unpublished draft, type yak, no matches. type yak1234ab no matches,.
-as dataverse admin, go to dashboard, manage users, search on yak, see all users. (this throws an exception at the moment, working that as a separate issue.

Attempting to remove roles using api endpoint with a key that does not have sufficient permissions gives incomplete/confusing error message: {"status":"ERROR","message":"Unable to revoke all roles: null"}

On manage users page, disabled user still appears, can make a superuser. Suggestion, make authentication type for disabled user, disabled.

Disable user if logged in, remains logged in and can edit account info. Update: this happens with delete user too. For disabled case, reenables it but without some stuff, for delete it adds a new account on save!

Can log in using builtin disabled user account, if that disabled user edited their account info in step 5

Can merge the same account into itself with result being user is deleted. In 5.3, throws constraint error, though could use better error msg.

Convert oauth (github) account link/page with same email as a newly created, enabled builtin account results in 500 error:

[2021-03-11T16:35:56.112+0000] [Payara 5.2020.6] [SEVERE] [] [javax.enterprise.resource.webcontainer.jsf.application] [tid: _ThreadID=92 _ThreadName=http-thread-pool::jk-connector(4)] [timeMillis: 1615480556112] [levelValue: 1000] [[
  Error Rendering View[/oauth2/convert.xhtml]
javax.el.ELException: /oauth2/convert.xhtml @23,91 value="#{OAuth2FirstLoginPage.welcomeMessage}": java.lang.NullPointerException

Caused by: java.lang.NullPointerException
	at edu.harvard.iq.dataverse.authorization.providers.oauth2.OAuth2FirstLoginPage.getWelcomeMessage(OAuth2FirstLoginPage.java:271)
Convert oauth (github) account with same email as disabled builtin account, sends user to convert account form, then when enter account to convert, provides both a success and failure message and oauth is logged in. Checking account status after shows merge didn't happen.
Convert shib account works with enabled builtin but on disabled builtin, gives both success and failure though it actually fails. Should check and prevent submitting account for conversion as ui above.

Test results:

1. Disable user endpoint doesn't care what permissions api key has but requires a key, otherwise http 405. Thought, based on discussion in pr and that it is an admin api, that it would not require a key.

I can't reproduce this and the automated tests do not use a token. However, the docs were incorrect and showed the use of an API token so I corrected them in abb8be0.

[Kevin] I see the doc was updated to remove the key. However, this is reproducible but after discussing with Gustavo, expected behavior for admin endpoints that assume console log on perms. So, not a bug but doc change removes confusion.

2. After disabling a user, some existing users cannot be found in assign roles ui, autocomplete.
As far as I can tell, autocomplete for assigning roles was completely broken in the branch, due to me botching the query. Fixed in 0efee28.

[Kevin] This is fixed.

3. Attempting to remove roles using api endpoint with a key that does not have sufficient permissions gives incomplete/confusing error message: {"status":"ERROR","message":"Unable to revoke all roles: null"}
The error wasn't being bubbled up from the command. Fixed in 3713104.

On manage users page, disabled user still appears, can make a superuser. Suggestion, make authentication type for disabled user, disabled.
As discussed in Slack, we're putting the word "disabled" in parentheses. In addition the JSON output from /api/admin/list-users now contained "deactivated" and "deactivatedTime". This code and the dashboard code are related. Fixed in c2b7b0a.

Disable user if logged in, remains logged in and can edit account info. Update: this happens with delete user too. For disabled case, reenables it but without some stuff, for delete it adds a new account on save!
Fixed in 3e59e70. Now the getUser code checks if the user in the session still exists and is not disabled.

Concern has been expressed that modifying getUser like this could result in a performance problem. A quick benchmark indicates that if you navigate to the homepage, sign up for an account, and then create a dataverse, the getUser method is called 264 times.

Can log in using builtin disabled user account, if that disabled user edited their account info in step 5
Fixed in 3e59e70, the same getUser fix mentioned above.

Can merge the same account into itself with result being user is deleted. In 5.3, throws constraint error, though could use better error msg.
Indeed, the user was deleted. This bug is probably preexisting in released versions. Fixed in 4f81fbd.

Convert oauth (github) account link/page with same email as a newly created, enabled builtin account results in 500 error:
I can't reproduce this. I'm on 4f81fbd.

Convert oauth (github) account with same email as disabled builtin account, sends user to convert account form, then when enter account to convert, provides both a success and failure message and oauth is logged in. Checking account status after shows merge didn't happen.
Fixed in fbded77 to prevent deactivated accounts from being converted to OAuth accounts.

Convert shib account works with enabled builtin but on disabled builtin, gives both success and failure though it actually fails. Should check and prevent submitting account for conversion as ui above.
Fix in fdc9cbe to prevent deactivated accounts from being converted to Shib accounts.

pdurbin commented:
> the superuser checkbox is still active

Fixed in c326af4c5. As shown in the screenshot below, the checkbox is disabled for deactivated users. I also enforced this in the underlying command as well as via API.

<img width="1417" alt="Screen Shot 2021-03-25 at 4 47 07 PM" src="https://user-images.githubusercontent.com/21006/112542956-89fd9500-8d8b-11eb-8b6e-25a9c91c54e4.png">

sekmiller commented:
Kevin, I put in an additional test if a user is trying to edit his account info while he is being deleted/deactivated. Hitting save will boot them to the root dv page, and if they try to log in again they will get the bad news.

scolapasta commented on src/main/resources/db/migration/V5.3.0.6__2419-deactivate-users.sql:
```diff
@@ -1,4 +1,6 @@
 -- Users can be deactivated.
 ALTER TABLE authenticateduser ADD COLUMN IF NOT EXISTS deactivated BOOLEAN;
+-- Prevent old users from having null for deactivated.
+UPDATE authenticateduser SET deactivated = false;
```
@pdurbin should there be any concern at all if this is run on a db that already has deactivated users (if someone removes this from their flyway table, for example)? Maybe not, but might be safer to add a where clause OR as a default on the Alter table, i.e. 
```
ALTER TABLE authenticateduser ADD COLUMN IF NOT EXISTS deactivated BOOLEAN DEFAULT false
```

sekmiller replied:
either works for me. I think "WHERE deactivated IS NULL;" stands out more as intent.

scolapasta commented on src/main/java/edu/harvard/iq/dataverse/authorization/providers/builtin/DataverseUserPage.java:
```diff
@@ -284,6 +284,12 @@ public void validateNewPassword(FacesContext context, UIComponent toValidate, Ob
+        
+        //First reget user to make sure they weren't deactivated or deleted
+        if (session.getUser().isAuthenticated() && !session.getUser(true).isAuthenticated()) {
+            return "dataverse.xhtml?alias=" + dataverseService.findRootDataverse().getAlias() + "&faces-redirect=true";
```
why not simplify and just go to homepage, i.e. "/"

sekmiller replied:
I can make that change. i was just using the pattern from the bottom of the save method.

sekmiller replied:
I'm going to leave it as is because when I tried return "/"; it left me in edit mode on the user page which I think would be confusing to the deleted/deactivated user.

scolapasta commented on src/main/java/edu/harvard/iq/dataverse/authorization/providers/builtin/DataverseUserPage.java:
```diff
@@ -284,6 +284,12 @@ public void validateNewPassword(FacesContext context, UIComponent toValidate, Ob
         boolean passwordChanged = false;
+        
+        //First reget user to make sure they weren't deactivated or deleted
+        if (session.getUser().isAuthenticated() && !session.getUser(true).isAuthenticated()) {
```
is the first part of this needed?

sekmiller replied:
I guess not. it's left over from when I did the re-get user within the save method.

sekmiller replied:
no. we need it because if we're creating an account the session user is a guest before they save the new account.
